src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2065,19 +2103,59 @@ GetInitialShardId(List *relationShardList)
 
 
 /*
- * TargetShardIntervalsForQuery performs shard pruning for all referenced relations
- * in the query and returns list of shards per relation. Shard pruning is done based
- * on provided restriction context per relation. The function sets multiShardQuery
- * to true if any of the relations pruned down to more than one active shard. It
- * also records pruned shard intervals in relation restriction context to be used
- * later on. Some queries may have contradiction clauses like 'and false' or
- * 'and 1=0', such queries are treated as if all of the shards of joining
- * relations are pruned out.
+ * TargetShardIntervalForFastPathQuery gets a query which is in
+ * the form defined by FastPathRouterQuery() and returns exactly
+ * one shard interval (see FastPathRouterQuery() for the detail).
+ *
+ * Also set the outgoing partition column value if requested via
+ * partitionValueConst
+ */
+static List *
+TargetShardIntervalForFastPathQuery(Query *query, Const **partitionValueConst,
+									bool *isMultiShardQuery)
+{
+	Const *queryPartitionValueConst = NULL;
+
+	Oid relationId = ExtractFirstDistributedTableId(query);
+	Node *quals = query->jointree->quals;
+
+	int relationIndex = 1;
+
+	List *prunedShardIntervalList =
+		PruneShards(relationId, relationIndex, make_ands_implicit((Expr *) quals),
+					&queryPartitionValueConst);
+
+	/* we're only expecting single shard from a single table */
+	Assert(FastPathRouterQuery(query));
+
+	if (list_length(prunedShardIntervalList) > 1)
+	{
+		*isMultiShardQuery = true;
+	}
+	else if (list_length(prunedShardIntervalList) == 1 &&
+			 queryPartitionValueConst != NULL)
```

Comment:
should this be `partitionValueConst != NULL`? meaning the caller asked for a partition column value?